app/queries/drug_stocks_query.rb

```diff
@@ -0,0 +1,103 @@
+class DrugStocksQuery
+  CACHE_VERSION = 1
+
+  def initialize(facilities:, for_end_of_month:)
+    @facilities = facilities
+    @for_end_of_month = for_end_of_month
+    # assuming that all facilities on the page have the same protocol
+    @protocol = @facilities.first.protocol
+    @state = @facilities.first.state
+  end
+
+  def protocol_drugs_by_category
+    @protocol_drugs_by_category ||= @protocol.protocol_drugs
+      .where(stock_tracked: true)
+      .order(:name, :dosage)
+      .sort_by(&:sort_key)
+      .group_by(&:drug_category)
+      .sort_by { |(drug_category, _)| drug_category }
```

Comment:
What does this sort achieve? Looks like we're making the final result a hash anyway, so does the ordering matter?

Reply:
This is the order in which the column groups, and columns are [displayed on the page](https://github.com/simpledotorg/simple-server/pull/2012/files/f6879780cc577fb49e4a5b74b17367da74c5da77#diff-d7918cfff1a26d2435b9e7511eb9fd48a0f1244a0106e31cd8d445c0b89d3204R22). We use this method as the canonical way to iterate over drug categories and drugs throughout this report as well.

The hash preserves the sorted order as per spec, I think. Can explicitly choose an ordered hash if we think that's better.

Reply:
Oh I see, the sort by drug category sorts by the enum order? Nice. I was under the impression that it would sort by category name, putting ARB drugs first.

Reply:
Oh it is ordered by name, and ARB is indeed first. We verified this with @dburka as well that this is okay.